rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,177 @@
+2. The client will call the `RegisterUsingToken` endpoint, specifying the name of the token, and providing the token that it has collected.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have allow-listed (RS[256, 384, 512])
+5. The server will check the `kid` of the JWT header, and obtain the relevant JWK from the cache or from the specified issuers well-known JWKS endpoint. It will then use the JWK to validate the token has been signed by the issuer.
+6. Other key claims of the JWT will be validated:
+  - Ensure the Issued At Time (iat) is in the past.
+  - Ensure the Expiry Time (exp) is in the future.
+7. The user's [configured Common Expression Language rule](#configuration) for the token will be evaluated against the claims, to ensure that the token is allowed to register with the Teleport cluster.
+8. Certificates will be generated for the client. The generated certificates will be non-renewable, as the client will proceed through the same steps to generate new certificates. This prevents exfiltratred credentials being used to repeatedly generate more credentials, maintaining access to the system.
+
+We will re-use `go-jose@v2` for validation of JWTs, since this library is already in use within Teleport.
+
+#### Caching JWKs
+
+Special attention should be given to the logic around caching the JWKs.
+
+We will cache these for two reasons:
+
+- Improve the performance of validating JWTs, as we will not need to make a HTTPS request to the issuer.
+- Improve the reliability, as we can validate JWTs even if the issuer is experiencing some downtime.
+- Reduce the impact of Teleport on an issuer. If onboarding a large number of nodes, we do not want to unduly place pressure on the issuer. 
+
+We should keep in mind the following considerations:
+
+- When we are presented with a JWT with a previously unseen `kid`, we should re-check the issuer's JWKs, as they may have begun issuing tokens with a new JWK.
+- We should ensure that the TTL of the cache entries is relatively short, as we want to allow an issuer to revoke a JWK that has been stolen.
+
+#### Configuration
+
+We will leverage the existing Token configuration kind as used by static tokens and IAM joining:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: my-gce-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: [Node]
+  join_method: oidc-jwt
+  issuer_url: https://accounts.google.com
+  allow: claims.aud == "noah.teleport.sh" && claims.google.compute_engine.project_id == "my-project" && claims.google.compute_engine.instance_name == "an-instance"
+```
+
+To allow the user to configure rules for what identities will be accepted, we will use the [Common Expression Language (CEL)](https://github.com/google/cel-spec). This allows a large degree of flexibility in the complexity of rules users can configure, but still allows simple expressions.
```

Comment:
Why can't we use the same rule engine that we use for RBAC?

Reply:
Hmm, I definitely see some merits in terms of consistency and reducing work in trying to re-use that.

I'm a little unsure how powerful the current RBAC rule engine is, so I'll need to look into that. As the structure of the claims in the JWT are out of our control, I'd hate to be in a situation where our language wasn't powerful enough for the user to construct the rules that would match the claims they want, and reject those that they dont.

Do we have any documentation available anywhere that explains what comparisons etc are available in the RBAC rule engine ? I can't seem to find any.

Reply:
https://goteleport.com/docs/setup/reference/predicate-language

Reply:
Having done some research into the Predicate Language in use in Teleport, it appears that this would be appropriate for our needs here, with extending to support some numeric values (at the moment, it is only possible to compare strings) or with some additional work to convert all claims to strings.